chalk-ir/src/lib.rs

```diff
@@ -1033,6 +1033,62 @@ impl<I: Interner> QuantifiedWhereClause<I> {
     }
 }
 
+#[derive(Clone, PartialEq, Eq, Hash, PartialOrd, Ord, HasInterner)]
+pub struct QuantifiedWhereClauses<I: Interner> {
+    interned: I::InternedQuantifiedWhereClauses,
+}
+
+impl<I: Interner> QuantifiedWhereClauses<I> {
+    pub fn new(interner: &I) -> Self {
+        Self::from(interner, None::<QuantifiedWhereClause<I>>)
+    }
+
+    pub fn interned(&self) -> &I::InternedQuantifiedWhereClauses {
+        &self.interned
+    }
+
+    pub fn from(
+        interner: &I,
+        clauses: impl IntoIterator<Item = impl CastTo<QuantifiedWhereClause<I>>>,
+    ) -> Self {
+        use crate::cast::Caster;
+        QuantifiedWhereClauses {
+            interned: I::intern_quantified_where_clauses(
+                interner,
+                clauses.into_iter().casted(interner),
+            ),
+        }
+    }
+
+    pub fn from_fallible<E>(
```

Comment:
This is getting to be an annoying amount of boilerplate -- not to be addressed in this PR, but I wonder if we can cleanup and encapsulate this pattern better.